### ISSUE TITLE
[FEA] Transition libcudf to a native C++ API w/ C wrapper API

### ISSUE THREAD

**Issue by jrhemstad** [rapidsai/cudf#1027]
**Is your feature request related to a problem? Please describe.**

Despite libcudf being a C++14 library, our native external API and types are all pure C. This is problematic because the "C"-ness of the API and types are viral and infect libcudf internals because the C constructs are not properly isolated from the internals of the library.

For example, `gdf_column` is a C-struct lacking any constructor, meaning that internal library devs are using clunky C APIs like `gdf_column_view` to create a `gdf_column` inside the library. Another example is `gdf_dtype` being a C enum rather than a proper strongly typed C++ enum. 

**Describe the solution you'd like**

The "C"-like elements of libcudf _internals_ should be completely removed and replaced with proper C++ constructs. Similarly, the native external API for libcudf should be C++ and be in the `cudf` namespace. My idea of the first few steps towards transitioning to a native C++ API are:

- All existing `gdf_*` C APIs would be transitioned to `cudf::*` C++ APIs
- Replace `gdf_column` with `cudf::column` class w/ appropriate constructors and member functions. It's member variables would mirror those of `gdf_column`
- Create `cudf::table` class as an abstraction for a collection of `cudf::column`s
- Transition to using exceptions in the C++ internals instead of requiring return error codes 

Furthermore, the monolithic `functions.h` is unwieldy. When transitioning to a C++ API, I propose breaking up our `include/` header files according to logical features/functionality, e.g., `cpp/include/join.hpp` would contain the APIs for Join`, `cpp/include/groupby.hpp` contains the API for groupby, etc. 

Finally, the two primary consumers of libcudf today: cuDF Python and BlazingSQL are capable of consuming the C++ API directly (cuDF Python via Cython and BlazingSQL natively in their C++ engine). Thus, focusing on our native C++ API in the short term does not alienate anyone. 

**Do not mistake my suggestion for completely removing the C API.** I am fully cognizant that an external C API is required for allowing some other languages to call into our library because the technology does not exist for them to call C++ functions. 

As such, our C API should be thin _wrappers_ on top of a native C++ API. All the wrappers would do is package and convert the C arguments and constructs into C++ constructs and then invoke the C++ API. The wrappers would also be responsible for converting an exception thrown from C++ into an appropriate error code returned by the C API. The reason for this is to provide a firewall against "C" constructs leaking into libcudf internals. 

I propose that the C wrappers live in `cudf/c` and the directory structure of `cudf/c` would mirror that of `cudf/cpp`, e.g., `cudf/c/src/join/join.cpp` would contain the implementation of the C wrappers for the Join API. Likewise `cudf/c/include/join.h` would have the `extern C` APIs.

**Comment by eyalroz** [comment]
I generally agree with Jacob.

That is, I believe cudf should have two APIs - in C++ and in C, either on "equal footing" or have the C++ API as the primary one and the C API considered "bindings". At a later (?) stage, the Python bindings could possibly be based on the C++ API (Cython supports this).

Now for some more detailed comments...:

> This is problematic because the "C"-ness of the API and types are viral and infect libcudf internals because the C constructs are not properly isolated from the internals of the library.

The "viral infection" is, IMHO, due to (at least) 6 factors:

1. No proper (intentional and explicit) isolation of the external C API from the bulk of the library. In particular, this means no underlying C++ API which the C API is a binding for. This encourages mixing up idioms between the two parts, and, obviously, you can only use C idioms in C++ rather than the other way around.
2. The implementation is itself not C++'ish enough. This might be a chicken-and-egg factor with the outcome of being C-infected, but once you write some fundamental innards in a C-ish way, you have to either have somewhat ugly adapters to keep the C-ishness at bay or it infects you from the inside-out, not just from the outside-in.
3. We only recently switched to C++14, and are only gradually adopting what that standard brings; and we haven't switched to C++17 on the host side (which we could do as far as CUDA is concerned - by isolating the kernel launches from non-C++14 code.) 
5. We are not using C++ "vocabulary types" from C++14 and later standard versions, which can be used even with earlier language standards using popular FOSS libraries: [std::variant](https://en.cppreference.com/w/cpp/utility/variant) and case classes, [std::optional](https://en.cppreference.com/w/cpp/utility/optional), [std::expected](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0323r4.html) for error handling, [std::span](https://en.cppreference.com/w/cpp/container/span), [std::string_view](https://en.cppreference.com/w/cpp/string/basic_string_view) and so on.
6. Some contributors have a lot of C experience, some C++98 experience, but medium or little "Modern C++" experience. Or, more exactly, many of us don't have "Modern C++ sensibilities", which would lead us to design things in a non-C++98-ish way. Others still have more such experience but don't feel confident enough to devise/suggest significant design changes in such a direction.

> The "C"-like elements of libcudf _internals_ should be completely removed and replaced with proper C++ constructs.

For those who are concerned with the "radicality" of this proposal: Not all of this has to be done at once. While eventually it would mean rewriting things down to the innermost definitions, we can start more superficially, then go deeper. That would mean more rewriting/refactoring work overall, but done over a longer period of time.

> * All existing `gdf_*` C APIs would be transitioned to `cudf::*` C++ APIs

Namespaces are great. Also, this is an example of a super-easy change that, doesn't really break anything. Of course - we should have intelligently-chosen sub-namespaces (e.g. perhaps for utility code, for kernels, for certain kinds of operators etc.)

> * Replace `gdf_column` with `cudf::column` class w/ appropriate constructors and member functions. It's member variables would mirror those of `gdf_column`

It's not just about making it a class. For example, I would not want for an "uninitialized column structure" to be possible. And - we would need to have clearer ownership semantics for the memory that a cudf::column points to. Perhaps we might also want an "owning column". All things to think about.

> * Transition to using exceptions in the C++ internals instead of requiring return error codes

Well, possibly. I've adopted this approach myself in my [C++ CUDA Runtime API Wrappers]( library, but exceptions have various drawbacks. We might not be very concerned with them, since our host-side code is not absolutely-all-the-way-optimized, but there is the `std::expected` way of doing things which should at least be considered. And there is [this proposal](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0709r2.pdf) of Herb Sutter, perhaps best illustrated by this fun CppCon 2018 talk: [What could possibly go wrong?](https://www.youtube.com/watch?v=GC4cp4U2f2E).


> Furthermore, the monolithic `functions.h` is unwieldy. When transitioning to a C++ API, I propose breaking up our `include/` header files according to logical features/functionality, e.g., `cpp/include/join.hpp` would contain the APIs for Join`, `cpp/include/groupby.hpp` contains the API for groupby, etc.

I'm in favor of doing that too, but frankly - that's almost orthogonal to the C-vs-C++ issue.

> I propose that the C wrappers live in `cudf/c` and the directory structure of `cudf/c` would mirror that of `cudf/cpp`, e.g., `cudf/c/src/join/join.cpp` would contain the implementation of the C wrappers for the Join API. Likewise `cudf/c/include/join.h` would have the `extern C` APIs.

So that's a pure "logistical" detail... seems basically fine to me. At this level of proposing things, I believe we should - already right now and irrespective of everything else - distinguish between C++ and C headers using different extensions, i.e. user `.hpp` for all C++-only header files.

**Comment by jrhemstad** [comment]
> That is, I believe `libcudf` should have three APIs - C++, C and Python, with the C and C++ ones either "on equal footing" or **have the C++ API as the primary one and the C API considered "bindings"**, like Python. 

Indeed, this is what I am suggesting. 

> 1. No proper (intentional and explicit) isolation of the external C API from the bulk of the library. In particular, this means no underlying C++ API which the C API is a binding for. This encourages mixing up idioms between the two parts, and, obviously, you can only use C idioms in C++ rather than the other way around.

Agreed, this is a chief motivation for providing the split between the primary C++ API and the wrapper C API, to enforce isolation. 

> 2. The implementation is itself not C++'ish enough. This might be a chicken-and-egg factor with the outcome of being C-infected, but once you write some fundamental innards in a C-ish way, you have to either have somewhat ugly adapters to keep the C-ishness at bay or it infects you from the inside-out, not just from the outside-in.

Agreed.

> 3. We only recently switched to C++14, and are only gradually adopting what that standard brings; and we haven't switched to C++17 on the host side (which we could do as far as CUDA is concerned - by isolating the kernel launches from non-C++14 code.)

Possibly, but I'd prefer to just wait to move to C++17 until `nvcc` supports it. 

> 4. We are not using C++ "vocabulary types" from C++14 and later standard versions, which can be used even with earlier language standards using popular FOSS libraries: [std::variant](https://en.cppreference.com/w/cpp/utility/variant) and case classes, [std::optional](https://en.cppreference.com/w/cpp/utility/optional), [std::expected](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0323r4.html) for error handling, [std::span](https://en.cppreference.com/w/cpp/container/span), [std::string_view](https://en.cppreference.com/w/cpp/string/basic_string_view) and so on.


> 5. Some contributors have a lot of C experience, some C++98 experience, but medium or little "Modern C++" experience. Or, more exactly, many of us don't have "Modern C++ sensibilities", which would lead us to design things in a non-C++98-ish way. Others still have more such experience but don't feel confident enough to devise/suggest significant design changes in such a direction.

Very much agree here. I'm hoping that a coding style guide can include some suggestions on "Modern C++ sensibilities" most relevant to libcudf development. 

> > * Replace `gdf_column` with `cudf::column` class w/ appropriate constructors and member functions. It's member variables would mirror those of `gdf_column`
> 
> It's not just about making it a class. For example, I would not want for an "uninitialized column structure" to be possible. And - we would need to have clearer ownership semantics for the memory that a cudf::column points to. Perhaps we might also want an "owning column". All things to think about.

One thing I neglected to think about or mention in the original issue is if we're moving to a C++ API, we should be possibly be using `arrow::Array` and `arrow::Table` instead of inventing our own classes and data structures. `libcudf` is after all supposed to be based on Arrow, we may as well use it natively. https://arrow.apache.org/docs/cpp/

> > Furthermore, the monolithic `functions.h` is unwieldy. When transitioning to a C++ API, I propose breaking up our `include/` header files according to logical features/functionality, e.g., `cpp/include/join.hpp` would contain the APIs for Join`, `cpp/include/groupby.hpp` contains the API for groupby, etc.
> 
> I'm in favor of doing that too, but frankly - that's almost orthogonal to the C-vs-C++ issue.

Indeed orthogonal, but since I'm suggesting refactoring much of the API, I wanted to mention where I see the file structure moving as well. 


> Most of these cannot be just "decided away", i.e. adopting Jacob's decision means investing some thought and then some time and effort in addressing each of these. (Not all of them immediately and at once though.)
> 
> > The "C"-like elements of libcudf _internals_ should be completely removed and replaced with proper C++ constructs.
> 
> For those who are concerned with the "radicality" of this proposal: Like I said above,Not all of this has to be done at once, on the level of the codebase just like on the level of our group orientation as developers. While eventually adopting this proposal would involve rewriting things down to the innermost definitions, we can start more superficially, then go deeper. That would mean more rewriting/refactoring work overall, but done over a longer period of time.

Definitely not all at once. Here's what I'm thinking for actionable steps in the right direction that we could start right away:

1. Create an initial `cudf::column` class that is a simple C++ analog to a `gdf_column` and `cudf::table` class for a set of `cudf::column`s
2. Provide a proof-of-concept of a native C++ API using simple `cudf::column` and `cudf::table` classes for a small set of features (I've already started this here in https://github.com/rapidsai/cudf/pull/577) 
3. Provide a C wrapper API for the C++ API from 1.
4. After 1, 2, 3 are well reviewed, accepted, and merged, they can serve as examples for all new features to provide a C++ API w/ C wrappers. 
5. As existing features are updated and refactored for various reasons, they should be updated to use the C++ API / C wrapper design.

**Comment by eyalroz** [comment]
About your suggested steps: Writing `cudf::column` properly is closer to the last thing we would be doing. It is not a simple class at all. So I'm not sure I would start that way.

About using `arrow::Array` or `arrow::Table`... [expletives remove here]. I wouldn't touch that with a 10-foot-pole if I didn't have to. To the extent that we need memory layout compatibility with it, or to take `arrow::Array` parameters - those are constraints I could live with, but certainly not adopt their types, functions and methods.

About waiting for nvcc to support C++17: Why wait? Even C++11 is not supported on the device-side, in terms of the standard library additions, and we're 8 years after it came out. So, we're talking about host-side code only for a lot of what we're going to write. And on the host side we're using the host compiler anyway.

**Comment by kkraus14** [comment]
@eyalroz I'm squashing this right now. We are not moving to C++17 because packaging and deploying it then becomes a support nightmare that we are not going to undertake at this time.

**Comment by jrhemstad** [comment]
> About your suggested steps: Writing `cudf::column` properly is closer to the last thing we would be doing. It is not a simple class at all. So I'm not sure I would start that way.

What I'm suggesting is providing the bare minimum C++ analog for a `gdf_column` such that we can start transitioning APIs to accept `cudf::column` parameters with minimal refactoring of libcudf internals and algorithms. Once the APIs are updated to accept `cudf::column`, we can always totally rewrite the internals without having to update the APIs.

**Comment by jrhemstad** [comment]
> About waiting for nvcc to support C++17: Why wait? Even C++11 is not supported on the device-side, in terms of the standard library additions, and we're 8 years after it came out. So, we're talking about host-side code only for a lot of what we're going to write. And on the host side we're using the host compiler anyway.

Aside from packaging issues, imagine you've got a class fully defined in a header. Some host member functions you may want to use C++17, but the class also has `__device__` member functions, meaning it has to be compiled with `nvcc`. Even though the host functions are compiled with the host compiler, `nvcc` will still refuse to compile the class. While there may be ways to cleverly separate these things, I think it would be more headache than anything else.

**Comment by felipeblazing** [comment]
For the modern stuff I agree guidelines are necessary particularly for fools like me. Though luckily our team is much more up to date on that stuff :).

> One thing I neglected to think about or mention in the original issue is if we're moving to a C++ API, we should be possibly be using `arrow::Array` and `arrow::Table` instead of inventing our own classes and data structures. `libcudf` is after all supposed to be based on Arrow, we may as well use it natively. https://arrow.apache.org/docs/cpp/

Ahhhh. We all knew this would happen someday but there are a few things that are going to make this tricky. So this will cause some problems for the stuff we have been doing for NVCategory and NVStrings becuase we are not following the same data layout and rely on those two classes to provide functionality in a way that is incompatible with the spirit of arrrow::array unless I am mistaken.  We could make a light wrapper for primitive types and arrow::array but will have issues with NVCategory and NVStrings unless we completely change our approach there. 

How would we handle these kinds of situations?

**Comment by eyalroz** [comment]
@kkraus14 :

> @eyalroz I'm squashing this right now. We are not moving to C++17 because packaging and deploying it then becomes a support nightmare that we are not going to undertake at this time.

Ok. However - we can definitely make use of some C++17 and C++20 _library_ goodies without switching to C++17 or C++20. All the examples I mentioned have widely-used and well-regarded C++14 (or earlier) implementations.

**Comment by jrhemstad** [comment]
> Ahhhh. We all knew this would happen someday but there are a few things that are going to make this tricky. So this will cause some problems for the stuff we have been doing for NVCategory and NVStrings becuase we are not following the same data layout and rely on those two classes to provide functionality in a way that is incompatible with the spirit of arrrow::array unless I am mistaken. We could make a light wrapper for primitive types and arrow::array but will have issues with NVCategory and NVStrings unless we completely change our approach there.
> 
> How would we handle these kinds of situations?

This is a really great point. I hadn't considered this how we're very non-standard wrt to Arrow and NVStrings/NVCategories.

To me this is just another nail in the coffin of using the Arrow types natively.

**Comment by datametrician** [comment]
We could work to align NVStrings/NVCategories to Apache Arrow String types

**Comment by jrhemstad** [comment]
> We could work to align NVStrings/NVCategories to Apache Arrow String types

If that's an option, then I suppose it's not a problem. 

(reposting this here)

One other issue is that we'd need some of the members of `arrow::Array` and `arrow::Table` to be `__device__` functions. Assuming we can make that happen, we've still got a problem: _relocatable device code_.  Since Arrow is an external shared library, in order to use `__device__` functions from the library, it would need to be compiled with relocatable device code enabled `-rdc=true`.

The problem is that using `rdc` means _all_ `__device__` functions in libcudf lose out on a whole host of optimizations because they have to be treated as indirect function calls, so no inlining, and every `__device__` function call would incur the overhead of having to follow the full function call ABI. So there could be performance regressions just by compiling in this mode.

If the Arrow classes we needed moved to header-only implementations, then this wouldn't be a problem.

**Comment by felipeblazing** [comment]
> The problem is that using `rdc` means _all_ `__device__` functions in libcudf lose out on a whole host of optimizations because they have to be treated as indirect function calls, so no inlining, and every `__device__` function call would incur the overhead of having to follow the full function call ABI. So there could be performance regressions just by compiling in this mode.

I dont think paying that cost is feasible. I am not sure if we can get header only implementations from this out of arrow. Might we not just have our own implementations that use the header file to know what to implement?

**Comment by jrhemstad** [comment]
> I dont think paying that cost is feasible. I am not sure if we can get header only implementations from this out of arrow. Might we not just have our own implementations that use the header file to know what to implement?

Do you mean maintaining our own header-only version of Arrow?

**Comment by jrhemstad** [comment]
Moving conversation related specifically to using the Arrow structures here: https://github.com/rapidsai/cudf/issues/1045

**Comment by rlratzel** [comment]
I like this proposal and would like to add something else to consider.

We could possibly generate the C API from the public C++ interfaces as part of the build process. This would prevent C constructs from leaking into the C++ core simply by not having them be part of the sources at all, and could eliminate the burden of having to maintain two APIs in parallel. Code generation also eliminates the chance of introducing bugs and inconsistencies due to simple oversights, and the generated C API would almost have to be thin by design too, which is another plus.

There are certainly nuances and idiomatic optimizations that the generator could have a hard time with, but the benefits might outweigh any special cases we'd have to discover and manually bake into the generator over time.

**Comment by jrhemstad** [comment]
Plenty of internal discussion has concluded that the focus for the foreseeable future will be on developing a pure C++ API and updating the internals accordingly. 

If or when we need a C API, we can create a new issue.
